libcloud/compute/drivers/kubevirt.py

```diff
@@ -316,69 +360,333 @@ def create_node(
+                    "'interface', 'name', 'network_type'."
+                )
+                raise KeyError(msg)
+        # add a default network
+        else:
+            interface = "masquerade"
+            network_name = "netw1"
+            network_type = "pod"
+
+        network_dict = {network_type: {}, "name": network_name}
+        interface_dict = {interface: {}, "name": network_name}
+
+        # ex_ports -> network.ports
+        ex_ports = ex_ports or {}
+        if ex_ports.get("ports_tcp"):
+            ports_to_expose = []
+            for port in ex_ports["ports_tcp"]:
+                ports_to_expose.append({"port": port, "protocol": "TCP"})
+            interface_dict[interface]["ports"] = ports_to_expose
+        if ex_ports.get("ports_udp"):
+            ports_to_expose = interface_dict[interface].get("ports", [])
+            for port in ex_ports.get("ports_udp"):
+                ports_to_expose.append({"port": port, "protocol": "UDP"})
+            interface_dict[interface]["ports"] = ports_to_expose
+
+        vm["spec"]["template"]["spec"]["networks"].append(network_dict)
+        vm["spec"]["template"]["spec"]["domain"]["devices"]["interfaces"].append(interface_dict)
+
+    @staticmethod
+    def _create_node_auth(vm, auth):
+        """
+        A part of create_node that deals with the authentication of the VM.
+        It will fill the vm with a cloud-init volume that injects the authentication.
+
+        :param vm: The VM template to be filled.
+        :param auth: The authentication method for the VM.
+                        See also the auth parameter in create_node.
+        :type auth: ``NodeAuthSSHKey`` or ``NodeAuthPassword``
+        :return: None
+        """
+        # auth requires cloud-init,
+        # and only one cloud-init volume is supported by kubevirt.
+        # So if both auth and cloud-init are provided, raise an error.
+        for volume in vm["spec"]["template"]["spec"]["volumes"]:
+            if "cloudInitNoCloud" in volume or "cloudInitConfigDrive" in volume:
+                raise ValueError(
+                    "Setting auth and cloudInit at the same time is not supported."
+                    "Use deploy_node() instead."
+                )
+
+        # cloud-init volume
+        cloud_init_volume = "auth-cloudinit-" + str(uuid.uuid4())
+        disk_dict = {"disk": {"bus": "virtio"}, "name": cloud_init_volume}
+        volume_dict = {
+            "name": cloud_init_volume,
+            "cloudInitNoCloud": {"userData": ""},
+        }
+
+        # cloud_init_config reference: https://kubevirt.io/user-guide/virtual_machines/startup_scripts/#injecting-ssh-keys-with-cloud-inits-cloud-config
+        if isinstance(auth, NodeAuthSSHKey):
+            public_key = auth.pubkey
+            cloud_init_config = (
+                """#cloud-config\n""" """ssh_authorized_keys:\n""" """  - {}\n"""
+            ).format(public_key)
```

Comment:
Could there be any surprises if `auth.pubkey` contains a leading or a trailing line break? Aka do we need to call `.strip()` on the value (ideally that would already happen in the base NodeAuth class, but I need to verify that is indeed the case).